joomla/joomla-cms
# [4] Ensure opcache file invalidate on file copy & delete

This is part 2 of https://github.com/joomla/joomla-cms/issues/32592 and directed at Joomla 4

### Summary of Changes

When PHP is configured to use modern Opcache, PHP will attempt to hold the compiled version cached in memory. 

There are many ways to configure opcache invalidations, including never revisiting the source PHP to compile it again ever (if [opcache.validate_timestamps](https://www.php.net/manual/en/opcache.configuration.php#ini.opcache.validate-timestamps) is disabled for example) 

To fully ensure that files that Joomla writes to the hard disk are taken into account by a server with a correct configured (or badly configured) opcache, we need to invalidate files from the opcache after writing a new version to the disk. 

**Joomla has previously used `opcache_reset` sporadically to do this, but  this is VERY BAD as causes a WHOLE SERVER cache clear - including other sites, and other sites you might not own! It can also cause spikes in CPU because all those sites need to recompile their opcodes!** 

### Testing Instructions

Hard to test unless you really know what you are doing and can reconfigure all your PHP stack to include opcaching. Also some of the edge cases this fixes will only become clear if you are an extension developer mass distributing extensions. 

Install some extensions - nothing should break. 

EDIT: Some more details on the testing can be found in comment https://github.com/joomla/joomla-cms/pull/32915#issuecomment-812562008

### Actual result BEFORE applying this Pull Request

You can install extensions

### Expected result AFTER applying this Pull Request

You can install extensions and opcache for each file is invalidated

### Documentation Changes Required

none

### Others

WordPress started this conversation 5 years ago and only last year got their changes into core 
https://core.trac.wordpress.org/changeset/48160


// cc @nikosdion
